light-clients/ics10-grandpa-cw/src/contract.rs

```diff
@@ -0,0 +1,361 @@
+pub const CONSENSUS_STATES_HEIGHTS: Map<Bytes, BTreeSet<Height>> =
+	Map::new("consensus_states_heights");
+pub const GRANDPA_HEADER_HASHES_STORAGE: Item<Vec<H256>> = Item::new("grandpa_header_hashes");
+pub const GRANDPA_HEADER_HASHES_SET_STORAGE: Map<Vec<u8>, ()> =
+	Map::new("grandpa_header_hashes_set");
+
+pub const GRANDPA_BLOCK_HASHES_CACHE_SIZE: usize = 500;
+
+#[derive(Clone, Copy, Debug, PartialEq, Default, Eq)]
+pub struct HostFunctions;
+
+impl light_client_common::HostFunctions for HostFunctions {
+	type BlakeTwo256 = BlakeTwo256;
+}
+
+impl grandpa_light_client_primitives::HostFunctions for HostFunctions {
+	type Header = RelayChainHeader;
+
+	fn ed25519_verify(
+		sig: &sp_core::ed25519::Signature,
+		msg: &[u8],
+		pub_key: &sp_core::ed25519::Public,
+	) -> bool {
+		use ed25519_zebra::{Signature, VerificationKey as PublicKey, VerificationKeyBytes};
+		let bytes: [u8; 64] = (sig.clone()).try_into().expect("signature is 64 bytes; qed");
+		let sig = Signature::from(bytes);
+		let slice: &[u8] = pub_key.as_ref();
+		let pub_key = PublicKey::try_from(
+			VerificationKeyBytes::try_from(slice).expect("key is always valid; qed"),
+		)
+		.expect("key is always valid; qed");
+		pub_key.verify(&sig, msg).is_ok()
+	}
+
+	fn insert_relay_header_hashes(_headers: &[<Self::Header as Header>::Hash]) {
+		// implementation of this method is in `Context`
+	}
+
+	fn contains_relay_header_hash(_hash: <Self::Header as Header>::Hash) -> bool {
+		// implementation of this method is in `Context`
+		true
+	}
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+	_deps: DepsMut,
+	_env: Env,
+	_info: MessageInfo,
+	_msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+	Ok(Response::default())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+	deps: DepsMut,
+	env: Env,
+	_info: MessageInfo,
+	msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+	let client = GrandpaClient::<HostFunctions>::default();
+	let mut ctx = Context::<HostFunctions>::new(deps, env);
+	let client_id = ClientId::from_str("08-wasm-0").expect("client id is valid");
+	let result = process_message(msg, client, &mut ctx, client_id.clone());
+	let data = match result {
+		Ok(res) => res,
+		Err(ContractError::Grandpa(e)) => to_binary(&ContractResult::error(e))?,
+		Err(e) => return Err(e),
+	};
+	let mut response = Response::default();
+	response.data = Some(data);
+	Ok(response)
+}
+
+fn process_message(
+	msg: ExecuteMsg,
+	client: GrandpaClient<HostFunctions>,
+	ctx: &mut Context<HostFunctions>,
+	client_id: ClientId,
+) -> Result<Binary, ContractError> {
+	// log!(ctx, "process_message: {:?}", msg);
+	let result = match msg {
+		ExecuteMsg::Status(StatusMsg {}) => {
+			let client_state = ctx
+				.client_state(&client_id)
+				.map_err(|e| ContractError::Grandpa(e.to_string()))?;
+			if client_state.frozen_height().is_some() {
+				Ok(to_binary("Frozen"))
+			} else {
+				Ok(to_binary("Active"))
```

Comment:
Don't we have `Expired` too?
https://github.com/cosmos/ibc-go/blob/main/modules/light-clients/07-tendermint/client_state.go#L94

Reply:
mmm can we?

Reply:
Sure, but we don't use it in Grandpa client anyway